### PR TITLE
Overhaul converting sampler finding code

### PR DESCRIPTION
**Pull request by trenouf**
This modifies the code that spots that a sampler is a YCbCr converting
sampler, allowing us to switch Builder's idea of the type of a sampler
back to v4i32.

Change-Id: I3d775746549553f38b331669d58e87d3eed3079c

### CODE REVIEW

**Comment by trenouf** [comment]
V2: Fixed a couple of bugs.

File: llpc/builder/llpcPipelineState.cpp [hunk 1]
```diff
@@ -804,6 +819,51 @@ void PipelineState::ReadUserDataNodes(
     m_userDataNodes = ArrayRef<ResourceNode>(m_allocUserDataNodes.get(), pNextOuterNode);
 }
 
+// =====================================================================================================================
+// Find the resource node for the given set,binding.
+// For nodeType == Unknown, the function finds any node of the given set,binding.
+// For nodeType == Resource, it matches Resource or CombinedTexture.
+// For nodeType == Sampler, it matches Sampler or CombinedTexture.
+// For nodeType == ResourceMappingNodeType::Sampler, it matches DescriptorSampler or DescriptorCombinedTexture.
```
Comment (nhaehnle) [inline llpc/builder/llpcPipelineState.cpp]: The spelling of the enum values here is inconsistent.

Comment (trenouf) [reply llpc/builder/llpcPipelineState.cpp]: Thanks; that was a mishap when I squashed commits from various things I was doing at the same time.

**Comment by trenouf** [comment]
> I think a commit message that elaborates _how_ the code that spots the YCbCr case is modified would be good.

Done.

**Comment by trenouf** [comment]
V3: Addressed Nicolai's comments.

File: llpc/builder/llpcPipelineState.h [hunk 2]
```diff
@@ -181,6 +181,14 @@ class PipelineState final : public Pipeline
     // Get user data nodes
     ArrayRef<ResourceNode> GetUserDataNodes() const { return m_userDataNodes; }
 
+    // Find the resource node for the given set,binding
+    std::pair<const ResourceNode*, const ResourceNode*> FindResourceNode(ResourceMappingNodeType  nodeType,
+                                                                         uint32_t                 descSet,
+                                                                         uint32_t                 binding) const;
+
+    // Get whether we have a converting sampler in the user data nodes.
+    bool GetHaveConvertingSampler() const { return m_haveConvertingSampler; }
```
Comment (amdrexu) [inline llpc/builder/llpcPipelineState.h]: GetHaveConvertingSampler -> HaveConvertingSampler
I think this is more natural.

File: llpc/builder/llpcPipelineState.cpp [hunk 3]
```diff
@@ -804,6 +819,54 @@ void PipelineState::ReadUserDataNodes(
     m_userDataNodes = ArrayRef<ResourceNode>(m_allocUserDataNodes.get(), pNextOuterNode);
 }
 
+// =====================================================================================================================
+// Find the resource node for the given set,binding.
```
Comment (amdrexu) [inline llpc/builder/llpcPipelineState.cpp]: set,binding -> {set,binding}

File: llpc/builder/llpcPipelineState.cpp [hunk 4]
```diff
@@ -804,6 +819,54 @@ void PipelineState::ReadUserDataNodes(
...
+// For nodeType == Unknown, the function finds any node of the given set,binding.
```
Comment (amdrexu) [inline llpc/builder/llpcPipelineState.cpp]: The same as above

File: llpc/builder/llpcPipelineState.cpp [hunk 5]
```diff
@@ -804,6 +819,54 @@ void PipelineState::ReadUserDataNodes(
...
+// For nodeType == Resource, it matches Resource or CombinedTexture.
+// For nodeType == Sampler, it matches Sampler or CombinedTexture.
+// For other nodeType, only a node of the specified type is returned.
+// Returns {topnode, node} where "node" is the found user data node, and "topnode" is the top-level user data
```
Comment (amdrexu) [inline llpc/builder/llpcPipelineState.cpp]: topnode -> topNode

**Comment by trenouf** [comment]
V4: Addressed Rex's comments.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8637738090239705bb2dbd9a568d6d5d6dba68d4
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev e642f608a62887d40d1f25509d2951a4a3576985</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev 80e5a4b11ad2058097e77746772ddc9ab2118e07</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b2b163fba8298b8ec6e36cee9666123d0069bfa8</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**